NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,410 @@
+					g_pSquirrel<context>->AsyncCall("NSHandleLoadResult", handle, "");
+					mutex.get().unlock();
+					return;
+				}
+
+				std::stringstream stringStream;
+				while (fileStr.peek() != EOF)
+					stringStream << (char)fileStr.get();
+
+				g_pSquirrel<context>->AsyncCall("NSHandleLoadResult", handle, stringStream.str());
+
+				fileStr.close();
+				mutex.get().unlock();
+				// side-note: this causes a leak?
+				// when a file is added to the map, it's never removed.
+				// no idea how to fix this - because we have no way to check if there are other threads waiting to use this file(?)
+				// tried to use m.try_lock(), but it's unreliable, it seems.
+			}
+			catch (std::exception ex)
+			{
+				spdlog::error("LOAD FAILED!");
+				mutex.get().unlock();
+				spdlog::error(ex.what());
+			}
+		});
+
+	readThread.detach();
+	return handle;
+}
+
+template <ScriptContext context> void SaveFileManager::DeleteFileAsync(fs::path file)
+{
+	// P.S. I don't like how we have to async delete calls but we do.
+	auto m = std::ref(mutexMap[file]);
+	std::thread deleteThread(
+		[m, file]()
+		{
+			try
+			{
+				m.get().lock();
+
+				fs::remove(file);
+
+				m.get().unlock();
+				// side-note: this causes a leak?
+				// when a file is added to the map, it's never removed.
+				// no idea how to fix this - because we have no way to check if there are other threads waiting to use this file(?)
+				// tried to use m.try_lock(), but it's unreliable, it seems.
+			}
+			catch (std::exception ex)
+			{
+				spdlog::error("DELETE FAILED!");
+				m.get().unlock();
+				spdlog::error(ex.what());
+			}
+		});
+
+	deleteThread.detach();
+}
+
+bool ContainsNonASCIIChars(std::string str)
+{
+	// we don't allow null characters either, even if they're ASCII characters because idk if people can
+	// use it to circumvent the file extension suffix.
+	return std::any_of(
+		str.begin(), str.end(), [](char c) { return static_cast<unsigned char>(c) > 127 || static_cast<unsigned char>(c) == '\0'; });
+}
+bool CheckFileName(std::string str)
+{
+	// we don't allow null characters either, even if they're ASCII characters because idk if people can
+	// use it to circumvent the file extension suffix.
+	return std::any_of(
+		str.begin(),
+		str.end(),
+		[](char c)
+		{
+			return static_cast<unsigned char>(c) > 127 || static_cast<unsigned char>(c) == '\0' || static_cast<unsigned char>(c) == '\\' ||
+				   static_cast<unsigned char>(c) == '//';
+		});
+}
+
+// void NSSaveFile( string file, string data )
+ADD_SQFUNC("void", NSSaveFile, "string file, string data", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm);
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (CheckFileName(fileName))
```

Comment:
You should check here that the path does not contain any `..`:

![image](https://user-images.githubusercontent.com/11993538/221383259-d979ba25-3c26-4edf-a873-e1bd3957fee6.png)

Reply:
Wait, but how did you do it without using / ?

Reply:
(I was a second too late with making new review xD)